resolwe/elastic/indices.py

```diff
@@ -0,0 +1,122 @@
+"""ElasticSearch indices."""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import operator
+import six
+
+import elasticsearch_dsl as dsl
+
+from guardian.models import GroupObjectPermission, UserObjectPermission
+
+from resolwe.flow.models import dict_dot
+
+
+class NoValue(object):
+    pass
+
+
+class BaseDocument(dsl.DocType):
+    users_with_permissions = dsl.String()
+    groups_with_permissions = dsl.String()
+
+
+class BaseIndex(object):
+    """Base index processor to be used in apps.
+
+    To work properly, subclass of this class must ovverride following methods
+    and attributes:
+
+      * object_type - class to which object must belong to be processed
+      * process_object - method that is called when object is created/updated
+
+    """
+
+    obj = None
+    object_type = None
+    mapping = {}
+    document_class = None
+    _preprocessed = False
+
+    def __init__(self, obj):
+        self.obj = obj
+
+    def filter(self):
+        pass
+
+    def preprocess_object(self):
+        pass
+
+    def generate_id(self):
+        object_type = self.object_type.__name__.lower()
+        index_name = self.document_class()._get_index()
+        return '{}_{}_{}'.format(object_type, index_name, self.obj.id)
+
+    def process_object(self):
+        if not self.object_type:
+            raise RuntimeError('`object_type` must be defined')
+
+        if not isinstance(self.obj, self.object_type):
+            return
+
+        if self.filter() is False:
+            return
+
+        if not self._preprocessed:
+            self.preprocess_object()
+            self._preprocessed = True
+
+        if not self.document_class:
+            raise RuntimeError('`document_class` must be defined')
+
+        document = self.document_class(meta={'id': self.generate_id()})
+
+        for field in document._doc_type.mapping:
+            if field in ['users_with_permissions', 'groups_with_permissions']:
+                continue  # this fields are handled separately
+
+            get_value_function = getattr(self, 'get_{}_value'.format(field), None)
+            if get_value_function:
+                setattr(document, field, get_value_function())
+                continue
+
+            if field in self.mapping:
+                if callable(self.mapping[field]):
+                    setattr(document, field, self.mapping[field]())
+                    continue
+
+                if isinstance(self.mapping[field], six.string_types):
+                    object_attr = dict_dot(self.obj, self.mapping[field], default=self.mapping[field])
```

Comment:
I don't think this is a good approach as typos will automatically change to default values (e.g. saying `fo.bar` instead of `foo.bar` will silently set all indexed values for that field to the string `'fo.bar'`). In order to set constants we should force them to be wrapped in lambda expressions or figure out a different way.